### PR TITLE
refactor(iroh-net): Introduce a minimal DerpUrl

### PR DESCRIPTION
**Pull request by flub**
## Description

This addresses two issues for now: some very minimal normalisation and
nicer Debug printing.

The type is purely a wrapper and does extremely minimal extra things
to the underlying Url.  There is no validation happening and no added
contraints to the URLs.  It is still possible to construct a DerpUrl
that will fail, e.g. by using the mailto: scheme.  The users of the
DerpUrl are expected to fail in such cases.

## Notes & open questions

I was initially trying to keep this crate-local to iroh-net.  However
iroh-net has a surprisingly large public API and that would be painful
and largely defeat the point to leave all those public APIs as Url.
Thus it is a public API.

As a consequence there are a few uses of this outside of the iroh-net
crate as well.  I think they are reasonable.

## Change checklist

- [x] Self-review.
- [x] Documentation updates if relevant.
- [x] Tests if relevant.

### CODE REVIEW

**Comment by flub** [comment]
See https://github.com/n0-computer/iroh/pull/1994 for the cargo-deny failure

**Review by rklaehn — APPROVED** [review]
I think there is some custom debug instance for AddrInfo that is no longer needed, assuming that AddrInfo also uses a DerpUrl.

```rust

impl Debug for AddrInfo {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("AddrInfo")
            .field("derp_url", &self.derp_url.as_ref().map(DD))
            .field("direct_addresses", &self.direct_addresses)
            .finish()
    }
}
```

File: iroh-net/src/derp.rs [hunk 1]
```diff
@@ -28,3 +36,109 @@ pub use self::server::{
...
+        Self(url)
+    }
+}
+
+impl FromStr for DerpUrl {
+    // Be aware, we are re-exporting another crate's public type in our API.
+    type Err = url::ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let inner = Url::from_str(s)?;
```
Comment (dignifiedquire) [inline iroh-net/src/derp.rs:78]: this doesn't do the domain ending check above, why?

File: iroh-net/src/derp.rs [hunk 2]
```diff
@@ -28,3 +36,109 @@ pub use self::server::{
...
+/// [`DerefMut`]: std::ops::DerefMut
+impl Deref for DerpUrl {
+    type Target = Url;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl fmt::Debug for DerpUrl {
```
Comment (dignifiedquire) [inline iroh-net/src/derp.rs:97]: missing `Display` impl for `DerpUrl`

File: iroh-net/src/derp.rs [hunk 3]
```diff
@@ -28,3 +36,109 @@ pub use self::server::{
     ClientConnHandler, MaybeTlsStream as MaybeTlsStreamServer, PacketForwarderHandler, Server,
 };
 pub use self::types::{MeshKey, PacketForwarder};
+
+/// A URL identifying a DERP server.
+///
+/// This is but a wrapper around [`Url`], with a few custom tweaks:
+///
+/// - A DERP URL is never a relative URL, so an implicit `.` is added at the end of the
```
Comment (dignifiedquire) [inline iroh-net/src/derp.rs:44]: this is never checked in the constructors

File: iroh-net/src/derp.rs (hunk 1 above)
Comment (flub) [reply iroh-net/src/derp.rs:78]: Oops, this was meant to use `DerpUrl::from` to make sure this check happened.

File: iroh-net/src/derp.rs (hunk 2 above)
Comment (flub) [reply iroh-net/src/derp.rs:97]: There is a derive `derive_more::Display`.  I've added a test for that functionality now as well.

File: iroh-net/src/derp.rs [hunk 4]
```diff
@@ -28,3 +36,109 @@ pub use self::server::{
...
+                url.set_host(Some(&domain)).ok();
+            }
+        }
+        Self(url)
+    }
+}
+
+impl FromStr for DerpUrl {
+    // Be aware, we are re-exporting another crate's public type in our API.
+    type Err = url::ParseError;
```
[the remark this answers is not in the document]
Comment (flub) [reply iroh-net/src/derp.rs:75]: We'd actively take away a lot of functionality though.  Admittedly you can get that back by constructing the URL manually first, so maybe that's fine.

Changed this.

File: iroh-net/src/derp.rs (hunk 2 above)
Comment (dignifiedquire) [reply iroh-net/src/derp.rs:101]: lol, I was just blind I guess

**Pull request merged**